File: parse_registry.py
import sys
import csv
from xml.parsers import expat
import codecs
import pandas as pd
from csv import writer

OUTPUT_FILE = "companies_founders.csv"


fields = ['NAME', 'SHORT_NAME', 'EDRPOU', 'ADDRESS', 'BOSS', 'KVED', 'STAN', "FOUNDER"]


def dump_company():
    global company
    for c in company:
        if company[c] == []:
            company[c] = ['']
    df = pd.DataFrame(company['FOUNDER'])
    for c in company:
        if c != 'FOUNDER':
            df[c] = company[c][0]
    df.to_csv(OUTPUT_FILE, mode = 'a', header = False, index = False)
        
  

def start_element(name, attrs):
    global company, content, isData
    if name in fields:
        isData = True
    else:
        isData = False

def company2nulls():
    global company
    company = {}
    for f in fields:
        company[f] = []

def end_element(name):
    global company, content
    if name == "RECORD":
        dump_company()
        company2nulls()
    elif name in fields:
        #print(content)
        company[name].append(content)
    content = ''


def char_data(data):
    global content
    #print(isData, data)
    if isData:
        content += data

if __name__ == "__main__":
    p = expat.ParserCreate()
    company2nulls()
    content = ''
    isData = False
    with open(OUTPUT_FILE, 'w') as of:
        of_writer = writer(of)
        of_writer.writerow(fields)
    p.StartElementHandler = start_element
    p.EndElementHandler = end_element
    p.CharacterDataHandler = char_data
    with codecs.open("/home/pavlo/texty.org.ua/registry/15.1-EX_XML_EDR_UO.xml", 'r', 'cp1251') as f:
        file_content = f.read()
        p.Parse(file_content)

File: parse_registry.R
library(XML)
library(stringr)

on_end <- function(name) {
  if (name == "FOUNDER") {
    list_from_xml[[name]] <<- c(list_from_xml[[name]], content_g)
  } else {
    if (name %in% headers) {
      list_from_xml[name] <<- content_g
    } else {
      if (name == "RECORD") {
        if (length(list_from_xml[['FOUNDER']]) == 0) {
          list_from_xml[['FOUNDER']] <- ''
        }
        df_temp <- data.frame(list_from_xml)
        xml_df <<- rbind(xml_df, df_temp)
        if (nrow(xml_df) >  10000) {
          filepath <- paste0(folder, paste0( as.character(count), ".csv"))
          write.csv(xml_df, filepath, row.names = FALSE)
          xml_df <<- data.frame()
          count <<- count + 1
        }
        list_from_xml[['FOUNDER']] <<- character()

      } else {
        if (name == "DATA") {
          filepath <- paste0(folder, paste0( as.character(count), ".csv"))
          write.csv(xml_df, filepath, row.names = FALSE)
        }
      }
    }
  }

}

gather_folder <- function(folder) {
  files <- list.files(folder)
  df <- data.frame()
  for (f in files) {
    print(f)
    df <- rbind(df, read.csv(paste0(folder,f)))
  }
  df
}

headers <- c('NAME', 'SHORT_NAME', 'EDRPOU', 'ADDRESS', 'BOSS', 'KVED', 'STAN', "FOUNDER")
filename <- "15.1-EX_XML_EDR_UO.xml"
folder <- "csvs/"
list_from_xml <- list()
for (h in headers) {
  list_from_xml[h] <- character()
}
file_number <- 1
count <- 1
el <- NULL
content_g <- ""
name_g <- ""
xml_df <- data.frame()
xmlEventParse(file = filename,
                    list(startElement = function(name, atts) {content_g <<- ''; name_g <<- name},
                         text = function(content){
                           if ((name_g) == "KVED") {
                             content_g <<- str_trim(paste(content_g, content))
                           } else {
                             content_g <<- str_trim(paste(content_g, content))
                           }
                         },
                         endElement = on_end))
write.csv(list_from_xml, paste0(  folder, as.character(file_number), ".csv"), row.names = FALSE)
print("gathering in single csv")
csv_from_xml  <- gather_folder(folder)
write.csv(csv_from_xml, paste0(filename, ".csv"))